File: sandbox/legacy-python/bootstrap/bootstrap/bootstrap_go/exec/set_env.go
// set_env: persist an environment variable that the desktop app reads
// on startup.
//
// Why not os.Setenv: we want the variable to survive past
// bootstrap.exe -- the desktop app launches as a separate process.
// Modifying machine-wide PATH/env is invasive and error-prone, so we
// instead write to a per-app config file at:
//
//   <bootDir>/runtime/bootstrap_env.json
//
// The desktop app reads this file at startup (via a tiny loader in
// core/supervisor or similar -- TBD on the Python side) and applies
// the values to its own os.environ. This keeps env mutation strictly
// inside OWLLM, with no side effects on the rest of the system.
//
// Args:
//   name   required -- env var name (e.g. "CUDA_HOME")
//   value  required -- string value
//   scope  optional -- "session" (default) or "user" or "machine".
//          Phase 0 only honors "session" (the JSON file). "user" and
//          "machine" are recorded but a follow-up will wire them
//          through to setx / HKCU registry entries.
package exec

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"

	"github.com/ruigro/LLM-Studio/LLM/bootstrap/bootstrap_go/plan"
)

const _envFileName = "bootstrap_env.json"

type SetEnvOpts struct {
	BootDir string
}

// SetEnv merges {name: value} into bootstrap_env.json under BootDir.
func SetEnv(opts SetEnvOpts, s plan.Step) error {
	name, err := argRequired(s.Args, "name")
	if err != nil {
		return fmt.Errorf("set_env: %w", err)
	}
	value, err := argRequired(s.Args, "value")
	if err != nil {
		return fmt.Errorf("set_env: %w", err)
	}
	scope, _ := argString(s.Args, "scope")
	if scope == "" {
		scope = "session"
	}
	if scope != "session" && scope != "user" && scope != "machine" {
		return fmt.Errorf("set_env: unknown scope %q", scope)
	}
	if scope != "session" {
		// Recorded but not yet applied -- fail loudly so the model
		// learns this avenue isn't ready, rather than silently
		// pretending to set machine-wide env.
		log.Printf("  set_env: scope=%q recorded only; "+
			"machine/user scope wiring not yet implemented", scope)
	}

	envPath := filepath.Join(opts.BootDir, "runtime", _envFileName)

	current := map[string]any{}
	if data, err := os.ReadFile(envPath); err == nil {
		_ = json.Unmarshal(data, &current)
	}

	// Stash under "session" / "user" / "machine" keys so the desktop
	// app can decide which buckets it honors.
	bucket, _ := current[scope].(map[string]any)
	if bucket == nil {
		bucket = map[string]any{}
	}
	bucket[name] = value
	current[scope] = bucket

	if err := os.MkdirAll(filepath.Dir(envPath), 0o755); err != nil {
		return fmt.Errorf("set_env: mkdir: %w", err)
	}
	out, err := json.MarshalIndent(current, "", "  ")
	if err != nil {
		return fmt.Errorf("set_env: marshal: %w", err)
	}
	if err := os.WriteFile(envPath, out, 0o644); err != nil {
		return fmt.Errorf("set_env: write: %w", err)
	}
	log.Printf("  set_env: %s=%q (scope=%s)", name, value, scope)
	return nil
}

File: python-app/bootstrap/bootstrap/bootstrap_rs/src/exec/set_env.rs
//! `set_env` — persist an environment variable that the desktop app
//! reads on startup.
//!
//! Port of `bootstrap_go/exec/set_env.go`. See the Go file's header
//! comment for the design rationale (why a JSON file instead of
//! `setx` / HKCU registry: isolation from system env). Behavior must
//! stay identical so the two implementations are drop-in
//! interchangeable.
//!
//! ## Args
//!
//! | key   | required | values                                  |
//! |-------|----------|-----------------------------------------|
//! | name  | yes      | env var name (e.g. `CUDA_HOME`)         |
//! | value | yes      | string value                            |
//! | scope | no       | `session` (default) / `user` / `machine`. Phase-0 only honors `session` |

use anyhow::{anyhow, Context, Result};
use serde_json::{Map, Value};
use std::fs;
use std::path::PathBuf;

use crate::args::{arg_required, arg_string};
use crate::plan::Step;

const ENV_FILE_NAME: &str = "bootstrap_env.json";

pub struct SetEnvOpts {
    pub boot_dir: PathBuf,
}

/// Merge `{name: value}` into `<boot_dir>/runtime/bootstrap_env.json`
/// under the requested scope bucket.
pub fn set_env(opts: &SetEnvOpts, step: &Step) -> Result<()> {
    let name = arg_required(&step.args, "name").context("set_env")?;
    let value = arg_required(&step.args, "value").context("set_env")?;

    let scope = arg_string(&step.args, "scope").unwrap_or_else(|| "session".to_string());
    match scope.as_str() {
        "session" | "user" | "machine" => {}
        other => return Err(anyhow!("set_env: unknown scope {:?}", other)),
    }
    if scope != "session" {
        // Recorded but not yet applied — log loudly so the model
        // learns this avenue isn't ready, matching the Go log line.
        eprintln!(
            "  set_env: scope={:?} recorded only; machine/user scope wiring not yet implemented",
            scope
        );
    }

    let env_path = opts.boot_dir.join("runtime").join(ENV_FILE_NAME);

    // Read existing file, if any. Tolerant: malformed JSON resets
    // to an empty map (matches the Go path's `_ = json.Unmarshal(...)`).
    let mut current: Map<String, Value> = match fs::read(&env_path) {
        Ok(bytes) => serde_json::from_slice(&bytes).unwrap_or_default(),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Map::new(),
        Err(e) => return Err(anyhow!("set_env: read {}: {}", env_path.display(), e)),
    };

    // Bucket = current["<scope>"] or new empty map.
    let mut bucket: Map<String, Value> = current
        .get(&scope)
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    bucket.insert(name.clone(), Value::String(value.clone()));
    current.insert(scope.clone(), Value::Object(bucket));

    if let Some(parent) = env_path.parent() {
        fs::create_dir_all(parent).context("set_env: mkdir")?;
    }

    let out =
        serde_json::to_vec_pretty(&Value::Object(current)).context("set_env: marshal")?;
    fs::write(&env_path, out).context("set_env: write")?;

    eprintln!("  set_env: {}={:?} (scope={})", name, value, scope);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use std::collections::BTreeMap;
    use tempfile::tempdir;

    fn read_env_file(boot_dir: &std::path::Path) -> BTreeMap<String, BTreeMap<String, Value>> {
        let p = boot_dir.join("runtime").join(ENV_FILE_NAME);
        let data = fs::read(&p).expect("env file exists");
        serde_json::from_slice(&data).expect("env file parses")
    }

    #[test]
    fn writes_session_bucket() {
        let tmp = tempdir().unwrap();
        let step = Step::new("set_env")
            .with_args(json!({ "name": "CUDA_HOME", "value": "/opt/cuda-12.1" }));
        set_env(
            &SetEnvOpts { boot_dir: tmp.path().to_path_buf() },
            &step,
        )
        .expect("set_env should succeed");
        let got = read_env_file(tmp.path());
        assert_eq!(
            got["session"]["CUDA_HOME"],
            Value::String("/opt/cuda-12.1".into())
        );
    }

    #[test]
    fn appends_to_existing_file() {
        let tmp = tempdir().unwrap();
        let opts = SetEnvOpts { boot_dir: tmp.path().to_path_buf() };
        let first = Step::new("set_env").with_args(json!({ "name": "A", "value": "1" }));
        let second = Step::new("set_env").with_args(json!({ "name": "B", "value": "2" }));
        set_env(&opts, &first).unwrap();
        set_env(&opts, &second).unwrap();
        let got = read_env_file(tmp.path());
        assert_eq!(got["session"]["A"], Value::String("1".into()));
        assert_eq!(got["session"]["B"], Value::String("2".into()));
    }

    #[test]
    fn rejects_unknown_scope() {
        let tmp = tempdir().unwrap();
        let step = Step::new("set_env")
            .with_args(json!({ "name": "X", "value": "y", "scope": "bogus" }));
        let err = set_env(
            &SetEnvOpts { boot_dir: tmp.path().to_path_buf() },
            &step,
        )
        .unwrap_err();
        assert!(err.to_string().contains("unknown scope"), "got: {}", err);
    }

    #[test]
    fn missing_name_fails() {
        let tmp = tempdir().unwrap();
        let step = Step::new("set_env").with_args(json!({ "value": "x" }));
        assert!(
            set_env(
                &SetEnvOpts { boot_dir: tmp.path().to_path_buf() },
                &step
            )
            .is_err()
        );
    }

    #[test]
    fn missing_value_fails() {
        let tmp = tempdir().unwrap();
        let step = Step::new("set_env").with_args(json!({ "name": "x" }));
        assert!(
            set_env(
                &SetEnvOpts { boot_dir: tmp.path().to_path_buf() },
                &step
            )
            .is_err()
        );
    }

    #[test]
    fn tolerates_corrupt_existing_file() {
        // The Go path drops a malformed JSON file silently and starts
        // fresh. Mirror that — otherwise a stray edit to bootstrap_env.json
        // would brick every subsequent set_env call.
        let tmp = tempdir().unwrap();
        let runtime = tmp.path().join("runtime");
        fs::create_dir_all(&runtime).unwrap();
        fs::write(runtime.join(ENV_FILE_NAME), b"not valid json {{{").unwrap();

        let step = Step::new("set_env").with_args(json!({ "name": "X", "value": "y" }));
        set_env(
            &SetEnvOpts { boot_dir: tmp.path().to_path_buf() },
            &step,
        )
        .expect("should recover from corrupt file");
        let got = read_env_file(tmp.path());
        assert_eq!(got["session"]["X"], Value::String("y".into()));
    }
}
